std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,629 @@
+        }
+    }
+
+    private bool insertNewNode()
+    {
+        static if (isShared)
+        import core.atomic : atomicOp;
+
+        void[] buf = parent.alignedAllocate(alignment, alignment);
+        if (buf is null)
+            return false;
+
+        auto localRoot = cast(AlignedBlockNode*) root;
+        auto newNode = cast(AlignedBlockNode*) buf;
+        ubyte[] payload = cast(ubyte[]) buf[AlignedBlockNode.sizeof .. $];
+        newNode.bAlloc = Allocator(payload);
+
+        newNode.next = localRoot;
+        newNode.prev = null;
+        if (localRoot)
+            localRoot.prev = newNode;
+        root = cast(typeof(root)) newNode;
+
+        static if (isShared)
+        {
+            atomicOp!"+="(numNodes, 1);
+        }
+        else
+        {
+            numNodes++;
+        }
+
+        return true;
+    }
+
+public:
+    static if (stateSize!ParentAllocator) ParentAllocator parent;
+    else alias parent = ParentAllocator.instance;
+
+    enum ulong alignment = theAlignment;
+
+    static if (hasMember!(ParentAllocator, "owns"))
+    Ternary owns(void[] b)
+    {
+        return parent.owns(b);
+    }
+
+    bool deallocate(void[] b)
+    {
+        static if (isShared)
+        import core.atomic : atomicOp;
+
+        if (b is null)
+            return true;
+
+        enum ulong mask = ~(alignment - 1);
+        // Round buffer to nearest `alignment` multiple to quickly find
+        // the 'parent' 'AlignedBlockNode'
+        ulong ptr = ((cast(ulong) b.ptr) & mask);
+        AlignedBlockNode *node = cast(AlignedBlockNode*) ptr;
+        if (node.bAlloc.deallocate(b))
+        {
+            static if (isShared)
+            {
+                atomicOp!"-="(node.bytesUsed, b.length);
+            }
+            else
+            {
+                node.bytesUsed -= b.length;
+            }
+            return true;
+        }
+        return false;
+    }
+
+    static if (hasMember!(ParentAllocator, "alignedAllocate"))
+    void[] allocate(size_t n)
+    {
+        static if (isShared)
+        import core.atomic : atomicOp, atomicLoad;
+
+        if (n == 0 || n > alignment)
+            return null;
+
+        static if (isShared)
+        {
+            lock.lock();
+            scope(exit) lock.unlock();
+        }
+
+        auto tmp = cast(AlignedBlockNode*) root;
+
+        // Iterate through list and find first node which has memory available
+        while (tmp)
+        {
+            auto next = tmp.next;
+            static if (isShared)
+            {
+                // Make sure nobody deletes this node while using it
+                tmp.keepAlive++;
+                if (next) next.keepAlive++;
+                lock.unlock();
+            }
+
+            auto result = tmp.bAlloc.allocate(n);
+            if (result.length == n)
+            {
+                static if (isShared)
+                {
+                    atomicOp!"+="(tmp.bytesUsed, n);
+                    lock.lock();
```

Comment:
You never unlock this

Reply:
A bit above I have a scope(exit) lock.unlock()